File: backend/routers/youtube.py
"""
/api/youtube — video, audio, thumbnail, subtitles, channel art
"""

import asyncio
import os
from fastapi import APIRouter, HTTPException, BackgroundTasks
from fastapi.responses import FileResponse
from pydantic import BaseModel
from typing import Optional

from utils import clean_yt_url
from services.youtube_service import (
    download_video,
    download_audio,
    download_thumbnail,
    download_subtitles,
    download_channel_art,
)

router = APIRouter()


class VideoRequest(BaseModel):
    url: str
    resolution: str = "best"
    format: str = "mp4"
    start_time: Optional[str] = None
    end_time: Optional[str] = None


class AudioRequest(BaseModel):
    url: str
    quality: str = "192"
    format: str = "mp3"


class SimpleRequest(BaseModel):
    url: str


class SubtitleRequest(BaseModel):
    url: str
    lang: str = "en"


class ChannelRequest(BaseModel):
    channel_url: str   # youtube.com/@handle  or  youtube.com/channel/UCxxxxxx


async def _delete_later(path: str, delay: int = 3600):
    await asyncio.sleep(delay)
    try:
        os.unlink(path)
    except Exception:
        pass


@router.post("/video")
async def video(req: VideoRequest, bg: BackgroundTasks):
    url = clean_yt_url(req.url.strip())
    try:
        path, filename = await download_video(
            url=url,
            resolution=req.resolution,
            fmt=req.format,
            start_time=req.start_time,
            end_time=req.end_time,
        )
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="video/mp4")


@router.post("/audio")
async def audio(req: AudioRequest, bg: BackgroundTasks):
    url = clean_yt_url(req.url.strip())
    try:
        path, filename = await download_audio(url=url, quality=req.quality, fmt=req.format)
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="audio/mpeg")


@router.post("/thumbnail")
async def thumbnail(req: SimpleRequest, bg: BackgroundTasks):
    url = clean_yt_url(req.url.strip())
    try:
        path, filename = await download_thumbnail(url)
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="image/jpeg")


@router.post("/subtitles")
async def subtitles(req: SubtitleRequest, bg: BackgroundTasks):
    url = clean_yt_url(req.url.strip())
    try:
        path, filename = await download_subtitles(url, req.lang)
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="text/plain")


@router.post("/channel-art")
async def channel_art(req: ChannelRequest, bg: BackgroundTasks):
    """
    Downloads avatar (JPG) and banner (JPG) for a YouTube channel.
    Returns JSON — frontend opens each asset URL to trigger download.

    {
      "assets": [
        {"type": "avatar", "filename": "...", "url": "/files/..."},
        {"type": "banner", "filename": "...", "url": "/files/..."}
      ]
    }
    """
    try:
        assets = await download_channel_art(req.channel_url.strip())
    except Exception as e:
        raise HTTPException(500, detail=str(e))

    for asset in assets:
        bg.add_task(_delete_later, f"downloads/{asset['filename']}")

    return {"assets": assets}

File: backend/src/routes/youtube.rs
use axum::extract::State;
use axum::response::Response;
use axum::Json;
use serde::Deserialize;
use serde_json::json;

use crate::config::AppConfig;
use crate::error::AppError;
use crate::services::youtube::YouTubeService;
use crate::utils::clean_yt_url;

#[derive(Deserialize)]
pub struct VideoRequest {
    pub url: String,
    #[serde(default = "default_resolution")]
    pub resolution: String,
    #[serde(default = "default_format")]
    pub format: String,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
}

fn default_resolution() -> String {
    "best".to_string()
}
fn default_format() -> String {
    "mp4".to_string()
}

#[derive(Deserialize)]
pub struct AudioRequest {
    pub url: String,
    #[serde(default = "default_quality")]
    pub quality: String,
    #[serde(default = "default_audio_format")]
    pub format: String,
}

fn default_quality() -> String {
    "192".to_string()
}
fn default_audio_format() -> String {
    "mp3".to_string()
}

#[derive(Deserialize)]
pub struct SimpleRequest {
    pub url: String,
}

#[derive(Deserialize)]
pub struct SubtitleRequest {
    pub url: String,
    #[serde(default = "default_lang")]
    pub lang: String,
}

fn default_lang() -> String {
    "en".to_string()
}

#[derive(Deserialize)]
pub struct ChannelRequest {
    pub channel_url: String,
}

pub async fn video(
    State(config): State<AppConfig>,
    Json(req): Json<VideoRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_yt_url(&req.url.trim());
    let (path, filename) = YouTubeService::download_video(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
        &req.resolution,
        &req.format,
        req.start_time.as_deref(),
        req.end_time.as_deref(),
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn audio(
    State(config): State<AppConfig>,
    Json(req): Json<AudioRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_yt_url(&req.url.trim());
    let (path, filename) = YouTubeService::download_audio(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
        &req.quality,
        &req.format,
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn thumbnail(
    State(config): State<AppConfig>,
    Json(req): Json<SimpleRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_yt_url(&req.url.trim());
    let (path, filename) = YouTubeService::download_thumbnail(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn subtitles(
    State(config): State<AppConfig>,
    Json(req): Json<SubtitleRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_yt_url(&req.url.trim());
    let (path, filename) = YouTubeService::download_subtitles(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
        &req.lang,
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn channel_art(
    State(config): State<AppConfig>,
    Json(req): Json<ChannelRequest>,
) -> Result<Json<serde_json::Value>, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let assets = YouTubeService::download_channel_art(
        &config.ytdlp_bin,
        &config.download_dir,
        req.channel_url.trim(),
    )
    .await?;

    Ok(Json(json!({ "assets": assets })))
}
